From 7aa147829de1dddcdacf416da358c7a4a9a94c29 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jonas=20Dre=C3=9Fler?= <verdre@v0yd.nl>
Date: Sat, 24 Oct 2020 13:20:26 +0200
Subject: [PATCH 1/3] clutter/actor: Add API to invalidate cached paint volumes

ClutterActors can override the get_paint_volume() vfunc in case they
draw outside the allocation. That's used by a bunch of actors, for
example ClutterText or StViewport in gnome-shell.

In case of StViewport, the paint volume returned depends on the value of
the StAdjustment, which means when we start to cache paint volumes more
agressively in ClutterActor, we'll need to add API that allows
StViewport to invalidate the paint volume. So introduce
clutter_actor_invalidate_paint_volume() to invalidate the cached paint
volume.

Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/1773>
---
 clutter/clutter/clutter-actor.c | 17 +++++++++++++++++
 clutter/clutter/clutter-actor.h |  3 +++
 2 files changed, 20 insertions(+)

diff --git a/clutter/clutter/clutter-actor.c b/clutter/clutter/clutter-actor.c
index 0a0884415d..7aac4ab847 100644
--- a/clutter/clutter/clutter-actor.c
+++ b/clutter/clutter/clutter-actor.c
@@ -19533,6 +19533,23 @@ clutter_actor_invalidate_transform (ClutterActor *self)
   transform_changed (self);
 }

+/**
+ * clutter_actor_invalidate_paint_volume:
+ * @self: A #ClutterActor
+ *
+ * Invalidates the cached paint volume of @self. This is needed for
+ * implementations overriding the #ClutterActorClass.get_paint_volume()
+ * virtual function and has to be called every time the paint volume
+ * returned by that function would change.
+ */
+void
+clutter_actor_invalidate_paint_volume (ClutterActor *self)
+{
+  g_return_if_fail (CLUTTER_IS_ACTOR (self));
+
+  self->priv->needs_paint_volume_update = TRUE;
+}
+
 gboolean
 clutter_actor_get_redraw_clip (ClutterActor       *self,
                                ClutterPaintVolume *dst_old_pv,
diff --git a/clutter/clutter/clutter-actor.h b/clutter/clutter/clutter-actor.h
index 39f96e2d5c..10b0a21748 100644
--- a/clutter/clutter/clutter-actor.h
+++ b/clutter/clutter/clutter-actor.h
@@ -931,6 +931,9 @@ GList * clutter_actor_peek_stage_views (ClutterActor *self);
 CLUTTER_EXPORT
 void clutter_actor_invalidate_transform (ClutterActor *self);

+CLUTTER_EXPORT
+void clutter_actor_invalidate_paint_volume (ClutterActor *self);
+
 G_END_DECLS

 #endif /* __CLUTTER_ACTOR_H__ */
--
GitLab


From 3d17e8d5908d46da422c140e9b08e473c999a154 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jonas=20Dre=C3=9Fler?= <verdre@v0yd.nl>
Date: Mon, 12 Oct 2020 10:58:38 +0200
Subject: [PATCH 2/3] clutter/actor: Properly invalidate cached paint volumes
 of actors

The priv->paint_volume field of ClutterActor stores the cached paint
volume in the actors local coordinate system. It consist of the actors
paint volume itself and the union of all children paint volumes.

We want to invalidate those cached paint volumes according to the
following rules:

- If an actors transformation matrix changes, all paint volumes of the
parent-tree need to be invalidated (that's because the parent-volumes
have unioned the actors paint volume). Our own paint volume does not
need invalidation since the transformation matrix is not applied to it.

- If an actors allocation-size changes, its own paint volume and all the
volumes of the parent-tree need to be invalidated. That's because the
allocation-size is used as the size of the paint volume.

- If a clip gets set or clip_to_allocation gets enabled for an actor,
its own paint volume and all the volumes of the parent-tree need to be
invalidated. That's because the clip is factored in when creating the
paint volume.

So far we did this invalidation in various places and the invalidation
up the parent-tree happened inside clutter_actor_real_queue_relayout().
We did not invalidate on changes to the actors transformation matrices
and the invalidation in clutter_actor_real_queue_relayout() was more
like a "big hammer" that probably invalidated unnecessarily a few times.

So introduce proper infrastructure to invalidate those cached paint
volumes of actors only in the cases where they actually need to be
invalidated. To do that, we reuse the transform_changed() function and
introduce a new function queue_update_paint_volume() that invalidates
the paint volumes up the actor tree.

Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/1773>
---
 clutter/clutter/clutter-actor.c | 57 +++++++++++++++++----------------
 1 file changed, 30 insertions(+), 27 deletions(-)

diff --git a/clutter/clutter/clutter-actor.c b/clutter/clutter/clutter-actor.c
index 7aac4ab847..956718dd86 100644
--- a/clutter/clutter/clutter-actor.c
+++ b/clutter/clutter/clutter-actor.c
@@ -1501,6 +1501,16 @@ queue_update_stage_views (ClutterActor *actor)
     }
 }

+static void
+queue_update_paint_volume (ClutterActor *actor)
+{
+  while (actor)
+    {
+      actor->priv->needs_paint_volume_update = TRUE;
+      actor = actor->priv->parent;
+    }
+}
+
 static void
 clutter_actor_real_map (ClutterActor *self)
 {
@@ -1516,8 +1526,6 @@ clutter_actor_real_map (ClutterActor *self)

   if (priv->unmapped_paint_branch_counter == 0)
     {
-      priv->needs_paint_volume_update = TRUE;
-
       /* We skip unmapped actors when updating the stage-views list, so if
        * an actors list got invalidated while it was unmapped make sure to
        * set priv->needs_update_stage_views to TRUE for all actors up the
@@ -2463,6 +2471,9 @@ transform_changed (ClutterActor *actor)
 {
   actor->priv->transform_valid = FALSE;

+  if (actor->priv->parent)
+    queue_update_paint_volume (actor->priv->parent);
+
   _clutter_actor_traverse (actor,
                            CLUTTER_ACTOR_TRAVERSE_DEPTH_FIRST,
                            absolute_geometry_changed_cb,
@@ -2492,7 +2503,7 @@ clutter_actor_set_allocation_internal (ClutterActor           *self,
 {
   ClutterActorPrivate *priv = self->priv;
   GObject *obj;
-  gboolean x1_changed, y1_changed, x2_changed, y2_changed;
+  gboolean origin_changed, size_changed;
   ClutterActorBox old_alloc = { 0, };

   g_return_if_fail (!isnan (box->x1) && !isnan (box->x2) &&
@@ -2504,10 +2515,11 @@ clutter_actor_set_allocation_internal (ClutterActor           *self,

   clutter_actor_store_old_geometry (self, &old_alloc);

-  x1_changed = priv->allocation.x1 != box->x1;
-  y1_changed = priv->allocation.y1 != box->y1;
-  x2_changed = priv->allocation.x2 != box->x2;
-  y2_changed = priv->allocation.y2 != box->y2;
+  origin_changed =
+    priv->allocation.x1 != box->x1 || priv->allocation.y1 != box->y1;
+  size_changed =
+    priv->allocation.x2 - priv->allocation.x1 != box->x2 - box->x1 ||
+    priv->allocation.y2 - priv->allocation.y1 != box->y2 - box->y1;

   priv->allocation = *box;

@@ -2516,10 +2528,7 @@ clutter_actor_set_allocation_internal (ClutterActor           *self,
   priv->needs_height_request = FALSE;
   priv->needs_allocation = FALSE;

-  if (x1_changed ||
-      y1_changed ||
-      x2_changed ||
-      y2_changed)
+  if (origin_changed || size_changed)
     {
       CLUTTER_NOTE (LAYOUT, "Allocation for '%s' changed",
                     _clutter_actor_get_debug_name (self));
@@ -2527,6 +2536,9 @@ clutter_actor_set_allocation_internal (ClutterActor           *self,
       /* This will also call absolute_geometry_changed() on the subtree */
       transform_changed (self);

+      if (size_changed)
+        queue_update_paint_volume (self);
+
       g_object_notify_by_pspec (obj, obj_props[PROP_ALLOCATION]);

       /* if the allocation changes, so does the content box */
@@ -2663,7 +2675,6 @@ clutter_actor_real_queue_relayout (ClutterActor *self)
   priv->needs_width_request  = TRUE;
   priv->needs_height_request = TRUE;
   priv->needs_allocation     = TRUE;
-  priv->needs_paint_volume_update = TRUE;

   /* reset the cached size requests */
   memset (priv->width_requests, 0,
@@ -2675,18 +2686,9 @@ clutter_actor_real_queue_relayout (ClutterActor *self)
   if (priv->parent != NULL)
     {
       if (priv->parent->flags & CLUTTER_ACTOR_NO_LAYOUT)
-        {
-          clutter_actor_queue_shallow_relayout (self);
-
-          /* The above might have invalidated the parent's paint volume if self
-           * has moved or resized. DnD seems to require this...
-           */
-          priv->parent->priv->needs_paint_volume_update = TRUE;
-        }
+        clutter_actor_queue_shallow_relayout (self);
       else
-        {
-          _clutter_actor_queue_only_relayout (priv->parent);
-        }
+        _clutter_actor_queue_only_relayout (priv->parent);
     }
 }

@@ -4757,6 +4759,7 @@ clutter_actor_set_clip_rect (ClutterActor          *self,
   else
     priv->has_clip = FALSE;

+  queue_update_paint_volume (self);
   clutter_actor_queue_redraw (self);

   g_object_notify_by_pspec (obj, obj_props[PROP_CLIP_RECT]);
@@ -9267,9 +9270,6 @@ clutter_actor_allocate (ClutterActor          *self,
       return;
     }

-  if (CLUTTER_ACTOR_IS_MAPPED (self))
-    self->priv->needs_paint_volume_update = TRUE;
-
   if (!origin_changed && !size_changed)
     {
       /* If the actor didn't move but needs_allocation is set, we just
@@ -11142,6 +11142,7 @@ clutter_actor_set_clip (ClutterActor *self,

   priv->has_clip = TRUE;

+  queue_update_paint_volume (self);
   clutter_actor_queue_redraw (self);

   g_object_notify_by_pspec (obj, obj_props[PROP_CLIP_RECT]);
@@ -11164,6 +11165,7 @@ clutter_actor_remove_clip (ClutterActor *self)

   self->priv->has_clip = FALSE;

+  queue_update_paint_volume (self);
   clutter_actor_queue_redraw (self);

   g_object_notify_by_pspec (G_OBJECT (self), obj_props[PROP_HAS_CLIP]);
@@ -15132,6 +15134,7 @@ clutter_actor_set_clip_to_allocation (ClutterActor *self,
     {
       priv->clip_to_allocation = clip_set;

+      queue_update_paint_volume (self);
       clutter_actor_queue_redraw (self);

       g_object_notify_by_pspec (G_OBJECT (self), obj_props[PROP_CLIP_TO_ALLOCATION]);
@@ -19547,7 +19550,7 @@ clutter_actor_invalidate_paint_volume (ClutterActor *self)
 {
   g_return_if_fail (CLUTTER_IS_ACTOR (self));

-  self->priv->needs_paint_volume_update = TRUE;
+  queue_update_paint_volume (self);
 }

 gboolean
--
GitLab


From 5a565b42586abb0c851888b827ffabc0cce38b75 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jonas=20Dre=C3=9Fler?= <verdre@v0yd.nl>
Date: Sat, 13 Mar 2021 20:43:13 +0100
Subject: [PATCH 3/3] clutter/actor: Update all last_paint_volumes before
 painting

Updating the last_paint_volume while painting has proven itself to be
quite prone to issues: First we had to make sure actors painted by
offscreen effects get their last_paint_volumes updated correctly (see
0320649a1c50993f308344e0dd296abaad0ee6d4), and now a new issue turned up
where we don't update the paint volumes while a fullscreen unredirect is
happening.

To stop those issues from happening and to lay the foundation for using
the last_paint_volume for other things, update the last_paint_volume in
a separate step before painting instead of doing it in
clutter_actor_paint().

To save some resources, avoid introducing another traversal of the
scenegraph and add that step into the existing step of updating the
stage_views lists of actors. To properly update the paint volumes, we
need to do that after finishing the queued redraws, which is why we move
clutter_stage_maybe_finish_queue_redraws() to happen before the new
clutter_stage_finish_layout().

Fixes https://gitlab.gnome.org/GNOME/mutter/-/issues/1699

Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/1773>
---
 clutter/clutter/clutter-actor-private.h |  4 +-
 clutter/clutter/clutter-actor.c         | 85 ++++++++-----------------
 clutter/clutter/clutter-stage-private.h |  2 +-
 clutter/clutter/clutter-stage-view.c    |  3 +-
 clutter/clutter/clutter-stage.c         |  7 +-
 5 files changed, 34 insertions(+), 67 deletions(-)

diff --git a/clutter/clutter/clutter-actor-private.h b/clutter/clutter/clutter-actor-private.h
index c34d772e87..46505777ae 100644
--- a/clutter/clutter/clutter-actor-private.h
+++ b/clutter/clutter/clutter-actor-private.h
@@ -260,8 +260,8 @@ float                           clutter_actor_get_real_resource_scale
 ClutterPaintNode *              clutter_actor_create_texture_paint_node                 (ClutterActor *self,
                                                                                          CoglTexture  *texture);
 
-void clutter_actor_update_stage_views (ClutterActor *self,
-                                       int           phase);
+void clutter_actor_finish_layout (ClutterActor *self,
+                                  int           phase);
 
 void clutter_actor_queue_immediate_relayout (ClutterActor *self);
 
diff --git a/clutter/clutter/clutter-actor.c b/clutter/clutter/clutter-actor.c
index 956718dd86..bcfffc714c 100644
--- a/clutter/clutter/clutter-actor.c
+++ b/clutter/clutter/clutter-actor.c
@@ -844,7 +844,6 @@ struct _ClutterActorPrivate
   guint needs_paint_volume_update   : 1;
   guint had_effects_on_last_paint_volume_update : 1;
   guint needs_update_stage_views    : 1;
-  guint children_painted            : 1;
 };
 
 enum
@@ -1495,7 +1494,7 @@ queue_update_stage_views (ClutterActor *actor)
       /* We don't really need to update the stage-views of the actors up the
        * hierarchy, we set the flag anyway though so we can avoid traversing
        * the whole scenegraph when looking for actors which need an update
-       * in clutter_actor_update_stage_views().
+       * in clutter_actor_finish_layout().
        */
       actor = actor->priv->parent;
     }
@@ -3589,30 +3588,6 @@ clutter_actor_paint_node (ClutterActor        *actor,
   return TRUE;
 }
 
-static void
-ensure_last_paint_volumes_updated (ClutterActor *self)
-{
-  ClutterActorPrivate *priv = self->priv;
-  ClutterActor *child;
-
-  /* Same entry checks as in clutter_actor_paint() */
-  if (CLUTTER_ACTOR_IN_DESTRUCTION (self))
-    return;
-
-  if (!CLUTTER_ACTOR_IS_TOPLEVEL (self) &&
-      ((priv->opacity_override >= 0) ?
-       priv->opacity_override : priv->opacity) == 0)
-    return;
-
-  if (!CLUTTER_ACTOR_IS_MAPPED (self))
-    return;
-
-  _clutter_actor_update_last_paint_volume (self);
-
-  for (child = priv->first_child; child; child = child->priv->next_sibling)
-    ensure_last_paint_volumes_updated (child);
-}
-
 /**
  * clutter_actor_paint:
  * @self: A #ClutterActor
@@ -3639,11 +3614,6 @@ clutter_actor_paint (ClutterActor        *self,
   g_autoptr (ClutterPaintNode) root_node = NULL;
   ClutterActorPrivate *priv;
   ClutterActorBox clip;
-  gboolean should_cull_out = (clutter_paint_debug_flags &
-                              (CLUTTER_DEBUG_DISABLE_CULLING |
-                               CLUTTER_DEBUG_DISABLE_CLIPPED_REDRAWS)) !=
-                             (CLUTTER_DEBUG_DISABLE_CULLING |
-                              CLUTTER_DEBUG_DISABLE_CLIPPED_REDRAWS);
   gboolean culling_inhibited;
   gboolean clip_set = FALSE;
 
@@ -3795,14 +3765,18 @@ clutter_actor_paint (ClutterActor        *self,
   if (!culling_inhibited && !in_clone_paint ())
     {
       gboolean success;
+      gboolean should_cull_out = (clutter_paint_debug_flags &
+                                  (CLUTTER_DEBUG_DISABLE_CULLING |
+                                   CLUTTER_DEBUG_DISABLE_CLIPPED_REDRAWS)) !=
+                                 (CLUTTER_DEBUG_DISABLE_CULLING |
+                                  CLUTTER_DEBUG_DISABLE_CLIPPED_REDRAWS);
       /* annoyingly gcc warns if uninitialized even though
        * the initialization is redundant :-( */
       ClutterCullResult result = CLUTTER_CULL_RESULT_IN;
 
-      if (should_cull_out)
-        _clutter_actor_update_last_paint_volume (self);
-
-      success = cull_actor (self, paint_context, &result);
+      success = should_cull_out
+        ? cull_actor (self, paint_context, &result)
+        : FALSE;
 
       if (G_UNLIKELY (clutter_paint_debug_flags & CLUTTER_DEBUG_REDRAWS))
         _clutter_actor_paint_cull_result (self, success, result, actor_node);
@@ -3819,22 +3793,8 @@ clutter_actor_paint (ClutterActor        *self,
   if (G_UNLIKELY (clutter_paint_debug_flags & CLUTTER_DEBUG_PAINT_VOLUMES))
     _clutter_actor_draw_paint_volume (self, actor_node);
 
-  priv->children_painted = FALSE;
-
   clutter_paint_node_paint (root_node, paint_context);
 
-  /* If an effect choose to not call clutter_actor_continue_paint()
-   * (for example offscreen effects might just paint their cached
-   * texture instead), the last_paint_volumes of the whole subtree
-   * still need to be updated to adjust for any changes to their
-   * eye-coordinates transformation matrices.
-   */
-  if (!priv->children_painted)
-    {
-      if (!culling_inhibited && !in_clone_paint () && should_cull_out)
-        ensure_last_paint_volumes_updated (self);
-    }
-
   /* If we make it here then the actor has run through a complete
      paint run including all the effects so it's no longer dirty */
   priv->is_dirty = FALSE;
@@ -3877,8 +3837,6 @@ clutter_actor_continue_paint (ClutterActor        *self,
       CoglFramebuffer *framebuffer;
       ClutterPaintNode *dummy;
 
-      priv->children_painted = TRUE;
-
       /* XXX - this will go away in 2.0, when we can get rid of this
        * stuff and switch to a pure retained render tree of PaintNodes
        * for the entire frame, starting from the Stage; the paint()
@@ -3954,6 +3912,11 @@ clutter_actor_pick (ClutterActor       *actor,
   ClutterActorBox clip;
   gboolean transform_pushed = FALSE;
   gboolean clip_set = FALSE;
+  gboolean should_cull = (clutter_paint_debug_flags &
+                          (CLUTTER_DEBUG_DISABLE_CULLING |
+                           CLUTTER_DEBUG_DISABLE_CLIPPED_REDRAWS)) !=
+                         (CLUTTER_DEBUG_DISABLE_CULLING |
+                          CLUTTER_DEBUG_DISABLE_CLIPPED_REDRAWS);
 
   if (CLUTTER_ACTOR_IN_DESTRUCTION (actor))
     return;
@@ -3969,7 +3932,7 @@ clutter_actor_pick (ClutterActor       *actor,
   /* mark that we are in the paint process */
   CLUTTER_SET_PRIVATE_FLAGS (actor, CLUTTER_IN_PICK);
 
-  if (priv->paint_volume_valid && priv->last_paint_volume_valid)
+  if (should_cull && priv->paint_volume_valid && priv->last_paint_volume_valid)
     {
       graphene_box_t box;
 
@@ -15871,8 +15834,8 @@ update_resource_scale (ClutterActor *self,
 }
 
 void
-clutter_actor_update_stage_views (ClutterActor *self,
-                                  gboolean      use_max_scale)
+clutter_actor_finish_layout (ClutterActor *self,
+                             gboolean      use_max_scale)
 {
   ClutterActorPrivate *priv = self->priv;
   ClutterActor *child;
@@ -15881,16 +15844,18 @@ clutter_actor_update_stage_views (ClutterActor *self,
       CLUTTER_ACTOR_IN_DESTRUCTION (self))
     return;
 
-  if (!priv->needs_update_stage_views)
-    return;
+  _clutter_actor_update_last_paint_volume (self);
 
-  update_stage_views (self);
-  update_resource_scale (self, use_max_scale);
+  if (priv->needs_update_stage_views)
+    {
+      update_stage_views (self);
+      update_resource_scale (self, use_max_scale);
 
-  priv->needs_update_stage_views = FALSE;
+      priv->needs_update_stage_views = FALSE;
+    }
 
   for (child = priv->first_child; child; child = child->priv->next_sibling)
-    clutter_actor_update_stage_views (child, use_max_scale);
+    clutter_actor_finish_layout (child, use_max_scale);
 }
 
 /**
diff --git a/clutter/clutter/clutter-stage-private.h b/clutter/clutter/clutter-stage-private.h
index 8f7ff31355..05412ace86 100644
--- a/clutter/clutter/clutter-stage-private.h
+++ b/clutter/clutter/clutter-stage-private.h
@@ -68,7 +68,7 @@ void                clutter_stage_maybe_finish_queue_redraws (ClutterStage
 GSList *            clutter_stage_find_updated_devices   (ClutterStage          *stage);
 void                clutter_stage_update_devices         (ClutterStage          *stage,
                                                           GSList                *devices);
-void                clutter_stage_update_actor_stage_views (ClutterStage        *stage);
+void                clutter_stage_finish_layout          (ClutterStage          *stage);
 
 CLUTTER_EXPORT
 void     _clutter_stage_queue_event                       (ClutterStage *stage,
diff --git a/clutter/clutter/clutter-stage-view.c b/clutter/clutter/clutter-stage-view.c
index 9b325e911c..b44bf88854 100644
--- a/clutter/clutter/clutter-stage-view.c
+++ b/clutter/clutter/clutter-stage-view.c
@@ -1171,9 +1171,10 @@ handle_frame_clock_frame (ClutterFrameClock *frame_clock,
   clutter_stage_emit_before_update (stage, view);
 
   clutter_stage_maybe_relayout (CLUTTER_ACTOR (stage));
-  clutter_stage_update_actor_stage_views (stage);
   clutter_stage_maybe_finish_queue_redraws (stage);
 
+  clutter_stage_finish_layout (stage);
+
   devices = clutter_stage_find_updated_devices (stage);
 
   frame = CLUTTER_FRAME_INIT;
diff --git a/clutter/clutter/clutter-stage.c b/clutter/clutter/clutter-stage.c
index 5b574b5e15..95e42e8589 100644
--- a/clutter/clutter/clutter-stage.c
+++ b/clutter/clutter/clutter-stage.c
@@ -930,7 +930,7 @@ clutter_stage_find_updated_devices (ClutterStage *stage)
 }
 
 void
-clutter_stage_update_actor_stage_views (ClutterStage *stage)
+clutter_stage_finish_layout (ClutterStage *stage)
 {
   ClutterActor *actor = CLUTTER_ACTOR (stage);
   ClutterStagePrivate *priv = stage->priv;
@@ -944,20 +944,21 @@ clutter_stage_update_actor_stage_views (ClutterStage *stage)
    * the paint.
    *
    * We're doing the whole thing twice and pass the phase to
-   * clutter_actor_update_stage_views() to allow actors to detect loops:
+   * clutter_actor_finish_layout() to allow actors to detect loops:
    * If the resource scale changes again after the relayout, the new
    * allocation of an actor probably moved the actor onto another stage
    * view, so if an actor sees phase == 1, it can choose a "final" scale.
    */
   for (phase = 0; phase < 2; phase++)
     {
-      clutter_actor_update_stage_views (actor, phase);
+      clutter_actor_finish_layout (actor, phase);
 
       if (!priv->actor_needs_immediate_relayout)
         break;
 
       priv->actor_needs_immediate_relayout = FALSE;
       clutter_stage_maybe_relayout (actor);
+      clutter_stage_maybe_finish_queue_redraws (stage);
     }
 
   g_warn_if_fail (!priv->actor_needs_immediate_relayout);
--
GitLab

